### ISSUE TITLE
4.12 upgrade breaks site mid-boot, so can't flush to fix.

### ISSUE THREAD

**Issue by NightJar** [silverstripe/silverstripe-framework#10721]
Bug:

https://github.com/silverstripe/silverstripe-framework/blob/f29577853a1fdcc1594f025a8779a44b95e8b192/src/Control/Session.php#L377-L378

Config can return anything. String, integer, boolean, or most likely, `null`.

But Cookie::validateSameSite expects a string, and only a string.

https://github.com/silverstripe/silverstripe-framework/blob/f29577853a1fdcc1594f025a8779a44b95e8b192/src/Control/Cookie.php#L110

@GuySartorelli tells me there's a new config setting as default for `cookie_samesite`.

https://github.com/silverstripe/silverstripe-framework/blob/4/src/Control/Session.php#L142

But that means a flush is necessary in order to pick that up, but after installing 4.12 it errors before any custom code is run (or we even reach flushing), meaning it's just 500's all the way back to 4.11.

```
[2023-03-08 13:35:54] error-log.ERROR: Uncaught Exception TypeError: "SilverStripe\Control\Cookie::validateSameSite(): Argument #1 ($sameSite) must be of type string, null given, 
called in vendor/silverstripe/framework/src/Control/Session.php on line 378" at vendor/silverstripe/framework/src/Control/Cookie.php line 110 {"excep
tion":"[object] (TypeError(code: 0): SilverStripe\\Control\\Cookie::validateSameSite(): Argument #1 ($sameSite) must be of type string, null given, called in vendor
/silverstripe/framework/src/Control/Session.php on line 378 at vendor/silverstripe/framework/src/Control/Cookie.php:110)"} []
```

recommend `$samesite = static::config()->get('cookie_samesite') ?? self::COOKIE_SAMESITE_DEFAULT` or some-such instead.

## Acceptance criteria
- Signature for `validateSameSite` stays the same.
- `$samesite = static::config()->get('cookie_samesite') ?? self::COOKIE_SAMESITE_DEFAULT` or some-such instead.
- Change is targeted at 4.12.

## PR
- https://github.com/silverstripe/silverstripe-framework/pull/10740

**Comment by michalkleiner** [comment]
or `Cookie::validateSameSite((string)$sameSite);`

**Comment by GuySartorelli** [comment]
`Cookie::validateSameSite();` throws an exception if the value is not one of the value values - which means passing `(string)null` will result in an exception being thrown.
The suggestion in the issue description would work though. @NightJar if you'd like to create a PR with that fix I'd be happy to review it.

**Comment by NightJar** [comment]
Fixes:

- #10740 
- https://github.com/silverstripe/silverstripe-cms/pull/2845

**Comment by emteknetnz** [comment]
I spend a fair trying to replicate this issue. In the end it seems you need the following:
- Have 4.11 installed
- Have SS_ENVIRONMENT_TYPE="live" (or probably "test", didn't try that though) - this is so that you need to be logged in to run ?flush=1
- Delete any session cookies from the browser
- Upgrade from 4.11 to 4.12, do not run dev/build flush=1 in CLI

Now when you attempt to ?flush=1 in the browser you'll be redirected to /Security/login as non-admins cannot flush when outside a "dev" environment type.  The login form will attempt to set a session cookie the mentioned config will be missing at this point

Seems like the issue can be easily worked around doing one of the following:
- Setting the environment mode on your local to "dev" before upgrading to 4.12 - I'm not sure why you wouldn't have "dev" mode enabled on your local at all times anyway
- Deleting the silverstripe-cache folder - normally located in /tmp/silverstripe-cache-php<version>-var-www
- Run dev/build flush=1 from CLI (no cookie required) - any deployment to non-dev environments should be doing this

I was initially concerned there may be some fragility with the config system and that we may need to add additional safeguards when adding new config variables as they may break upgrades, but after investigating this I'm happy that there's enough robustness in the config system

**Comment by GuySartorelli** [comment]
It's worth noting that CLI can sometimes have a _separate_ cache from the webserver (e.g. if a different PHP binary is used, or it is run as a different user - which in production is likely, as the www-data user for apache is not likely to be the same as the user running sake commands via the CLI).

The workaround also doesn't work for production environments if:
- User can't delete the cache directly (no ssh access)
- User can't change the environment mode (no ssh access)

I think high impact is still warranted for this issue.

**Comment by michalkleiner** [comment]
I also agree with the high impact tag as this is an issue mainly in a production environment if someone upgrades from 4.11 to 4.12 which is still the latest and has been for some time.

**Comment by emteknetnz** [comment]
> It's worth noting that CLI can sometimes have a separate cache from the webserver

Right so the dev/build flush=1 in your deployment pipeline won't actually flush for the webserver cache. Yeah, that's not ideal, I guess that'll happen in the real-world and it's easy to miss when validating that your deployment pipeline is working as it should. It is technically a server config issue and not an application issue, though I guess that's not very pragmatic

Given that, maybe there is some fragility in the config system? IMO You should never need to run ?flush=1 via the browser on a prod system. 

We could do something similar to https://github.com/silverstripe/silverstripe-graphql/pull/530 i.e. rebuild the whole cache on miss with a minimum interval of 15 minutes, though this approach makes less sense that in the graphql example because a developer could have simply mistyped a config variable name they want to get that'll never be there and it'll just keep flushing for no reason

Maybe it could work if we allow a miss, rebuild, 15 minutes later if the same thing missed then throw a fatal error?

**Comment by GuySartorelli** [comment]
We only did that in graphql because we couldn't reproduce the actual underlying problem in order to diagnose and actually fix it. I think adding something like that into the config system would be adding a pretty complex bandaid where it's not really warranted. The only case where a config miss has caused _real_ problems that I'm aware of is this one specific instance. The immediate fix is something like https://github.com/silverstripe/silverstripe-framework/pull/10740 (without the unrelated changes).

If there is a deeper underlying problem to fix, it is one of these:
1. It is possible for configuration to be required before a flush event occurs in production environments.
    - if we think this is a problem, we should look deeper into the initial lifecycle events of booting up manifests etc and see if there's something we can do there to avoid config cache misses before config can be flushed
    - I think in this case it's somewhat unavoidable though, since you need to authenticate to flush and authentication requires cookies
    - Maybe there's something we can do to provide a cookie-less authentication method specifically for scenarios like this but I think it's overkill.
1. Configuration that has never been defined returns `null` without checking if a new configuration property was defined post-flush
    - I don't think there's any reasonably performant way to handle this - I think cache miss should always return `null`. Otherwise the cache has diminished value.
    - This is mitigated somewhat with the `SS_FLUSH_ON_DEPLOY` environment variable - but only if the deployer has requisite access to set that environment variable.

All of that said, I think this is an unusual edge case which happened because the new configuration is required for setting cookies which happens on every request, which prevents authenticating in order to perform a flush.

At best I think we could look at having scenarios like this one be part of our regression testing process.

**Comment by emteknetnz** [comment]
> since you need to authenticate to flush and authentication requires cookies

That's true, though I don't think it's entirely true. If you delete the /tmp/silverstripe-cache where SS_ENVIRONMENT_TYPE="live" and visit the website unauthenticated it'll rebuild the cache, the same as flush would. My suggestion to flush on cache missing is effectively the same thing, you do an authenticated flush that can only be triggered by a cache miss.  The difference is that normally flush on cache miss when there is no cache, this would extend that to flush on cache miss when there is a cache

> I don't think there's any reasonably performant way to handle this - I think cache miss should always return null. Otherwise the cache has diminished value.

Yeah there is no performant way to handle this, due to the way that the silverstripe is an all-or-nothing cache that cannot handle cache misses.  Not sue so sure I agree with cache misses returning null though. For a more "normal" cache, say an in memory cache in php where you just save values to a private array with a cache key, if there's a cache miss you'd simply run some code and then save the result to the private array along with a cache key.  However in Silverstripe there's no way to "just run some code to calculate one value" this since working out config involves tranversing all the of the yml files and all of the private static things.

Of course the "normal" cache also has the assumption that the cache key that you're using is sensical (it's usually auto generated), which gets to my point about rebuilding on miss is bad when a developer mistypes a cache key

> pretty complex bandaid where it's not really warranted

Yup this sort of solution is certainly a debatable one.  For this particular instance it's obviously total overkill since we can easily solve it with a code fix.  I'm thinking more in terms of unidentified issues, both now and in the future, that could happen on upgrade due to a stale cache that doesn't have an easy way to update itself.

> If there is a deeper underlying problem to fix, it is one of these: .. It is possible for configuration to be required before a flush event occurs in production environments. 

I'm more concerned that flush events aren't happening at all

**Comment by michalkleiner** [comment]
> if you delete the /tmp/silverstripe-cache where SS_ENVIRONMENT_TYPE="live"

That's the main problem. There are live environments, and probably more than we think, where you don't have access to manually delete stuff in the /tmp/ folder.

**Comment by NightJar** [comment]
My environment is `dev`.

**Comment by NightJar** [comment]
> I'm more concerned that flush events aren't happening at all

[This is not the issue](https://github.com/silverstripe/silverstripe-framework/blob/c278576996f73cf8be53924c0fe55dff86b5045e/_config/requestprocessors.yml#L11-L13). As per the issue description:

> errors before [...] we even reach flushing

Work arounds are not a fix. _Especially when undocumented._
The config system is not the problem.
Blaming the user is unhelpful. The system worked, then a minor upgrade stopped it working. This is a system issue.
Deployment processes are not the concern of the system.
There is no CLI conditional logic on sessions. A session start is always attempted (triggering the error).
Flushes will happen if the problem is resolved.

Focusing on areas outside the problem is just a distraction that slows down the remedial process.

We know the problem. We know the cause. We have a fix. Considering systemic impact is good. Overthinking it isn't ;)

**Comment by emteknetnz** [comment]
Linked PR has been merged and released as [4.12.4](https://github.com/silverstripe/silverstripe-framework/releases/tag/4.12.4)
